### PR TITLE
Increase coverage

### PR DESCRIPTION
**Pull request by ahartikainen**
Now mostly everything (excluding tfp + pymc3) that is not really hard to impossible to do is tested.

This pr does a bunch of changes.

### CODE REVIEW

File: arviz/tests/test_stats.py [hunk 1]
```diff
@@ -113,8 +113,9 @@ def test_summary_fmt(centered_eight, fmt):
     assert summary(centered_eight, fmt=fmt) is not None
 
 
-def test_summary_stat_func(centered_eight):
-    assert summary(centered_eight, stat_funcs=[np.var]) is not None
+# Issue with xarray.apply_ufunc
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_stats.py]: Yeah, that sounds reasonable.

You can make a commit when you have time

[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_stats.py]: I added `@pytest.mark.xfail` for tests that are failing due to bugs. Thanks for the tip.

[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/tests/test_stats.py]: If you catch new ones can you add a short snippet of what is failing as well? I can go back and do it too if you're short on time

The reason I ask is this will help others when they try to go back and fix the bug. This way when people come to contribute they both have a test and comment explaining what to look into!

Examples are here
https://docs.pytest.org/en/latest/skipping.html#examples

[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_stats.py]: Feel free to edit. I can do this later (3 hours) but if you have time, you can do it.

File: arviz/plots/forestplot.py [hunk 2]
```diff
@@ -372,10 +371,7 @@ def make_bands(self, ax):
                 prev_color_index = self.colors.index(color)
                 y_prev = y
 
-        if plotter is None:
-            offset = 1
```
Comment (ahartikainen) [inline arviz/plots/forestplot.py]: Given that if this is True, the `y_prev` would be `None` which would raise an Exception.

File: arviz/stats/diagnostics.py [hunk 3]
```diff
@@ -316,10 +316,7 @@ def geweke(values, first=0.1, last=0.5, intervals=20):
 
         zscores.append([start, z_score])
 
-    if intervals is None:
-        return np.array(zscores[0])
-    else:
-        return np.array(zscores)
```
Comment (ahartikainen) [inline arviz/stats/diagnostics.py]: intervals can not be None

File: arviz/tests/helpers.py [hunk 4]
```diff
@@ -307,34 +309,21 @@ def load_cached_models(eight_school_params, draws, chains):
...
-        nchain = fit.sim["chains"]
-        extract = {}
-        for key, ary in extract_permuted.items():
-            ary = np.asarray(ary)[ary_order]
-            if ary.shape:
-                ary_shape = ary.shape[1:]
-            else:
-                ary_shape = ary.shape
-            ary = ary.reshape((-1, nchain, *ary_shape), order="F")
-            extract[key] = ary
```
Comment (ahartikainen) [inline arviz/tests/helpers.py]: This is adhoc code for PyStan 2.17. Testing should be done 2.18+. Let's not test anymore with 2.17 or earlier.

File: arviz/tests/test_plot_utils.py [hunk 5]
```diff
@@ -40,8 +50,6 @@ def test_dataset_to_numpy_combined(sample_dataset):
     var_names, data = xarray_to_ndarray(data, combined=True)
 
     assert len(var_names) == 2
-    if var_names[0] == "tau":
-        data = data[::-1]
```
Comment (ahartikainen) [inline arviz/tests/test_plot_utils.py]: remove code, because this is already fixed.

File: arviz/tests/test_plots.py [hunk 6]
```diff
@@ -55,19 +55,7 @@ def clean_plots(request, save_figs):
...
-                try:
-                    plt.savefig("{0}.png".format(os.path.join(save_figs, request.node.name)))
-
-                except Exception as err:  # pylint: disable=broad-except
-                    if i == 2:
-                        raise err
-                    time.sleep(0.250)
-                else:
-                    break
-
```
Comment (ahartikainen) [inline arviz/tests/test_plots.py]: I don't think this retry loop was really needed for anything. I remember we had a discussion concerning this, but still, I feel like this is not needed. The error was something else.

is not needed == doesn't actually fix any problems, because even if the user has the image open, closing them in 0.75 s is just overengineering on our part.

**Comment by canyon289** [comment]
Let's delete it. I can do it in another pr

File: arviz/tests/test_plots.py (hunk 6 above)
Comment (ahartikainen) [reply arviz/tests/test_plots.py]: If you want to do it in another PR you can. I can merge it with master.

**Comment by canyon289** [comment]
@ahartikainen It looks like I caused you a merge conflict with my latest pr. 

In general do you think its worth merging your branch as is before anymore merge conflicts come up?

**Comment by ahartikainen** [comment]
Let's not merge anything from this branch. I will fix that.

Also I think some of these changes should go other PR, but I'm not finished yet.

**Comment by ahartikainen** [comment]
Okey, I think this ready for review.

Please check especially all non--test code that I changed.

File: arviz/tests/test_diagnostics.py [hunk 7]
```diff
@@ -56,3 +70,25 @@ def test_geweke(self):
...
+            geweke(np.random.randn(10), first=0)
+        # upper bound
+        with pytest.raises(ValueError):
+            geweke(np.random.randn(10), last=1)
+        # sum larger than 1
+        with pytest.raises(ValueError):
+            geweke(np.random.randn(10), first=0.9, last=0.9)
+
+    def test_ks_summary(self):
+        #  fake
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_diagnostics.py]: It probably be more verbose. Means that khat values are just fake and not from the psislw function

File: .travis.yml [hunk 8]
```diff
@@ -64,6 +65,12 @@ script:
                             "source /root/.bashrc && ./scripts/lint.sh";
     fi
 
+  # Tests saving plots and deleting files
+  - if [[ $NAME == LINT ]]; then docker run  --mount type=bind,source="$(pwd)",target=/opt/arviz --name arviz_test arviz:latest /bin/bash -c \
+                                              "source /root/.bashrc && pytest -v arviz/tests/ --save -k plot_forest && pytest -v arviz/tests/ --save -k plot_forest --cov=arviz/";
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply .travis.yml]: First pass creates figures and saves them, second pass also needs to clear the folder.

--cov is runned only once, so we don't need to combine any files later.

File: arviz/tests/test_plot_utils.py [hunk 9]
```diff
@@ -22,6 +22,16 @@ def sample_dataset():
     return mu, tau, data
 
 
+# test code that is hard to reach
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_plot_utils.py]: Sure we can make it a docstring

File: arviz/tests/test_plot_utils.py [hunk 10]
```diff
@@ -22,6 +22,16 @@ def sample_dataset():
...
+def test_make_2d():
+    assert len(make_2d(np.array([2, 3, 4])).shape) == 2
+
+
+# test code that is hard to reach
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_plot_utils.py]: Yep

File: arviz/tests/test_plot_utils.py [hunk 11]
```diff
@@ -68,6 +76,14 @@ def test_xarray_var_iter_ordering_uncombined(sample_dataset):  # pylint: disable
         ]
 
 
+# test code that is hard to reach
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_plot_utils.py]: Yep

File: arviz/tests/test_plots.py [hunk 12]
```diff
@@ -226,6 +252,16 @@ def test_plot_joint_discrete(discrete_model):
     assert axjoin
 
 
+@pytest.mark.parametrize("model_fit", ["pymc3_fit", "stan_fit", "tfp_fit"])
+def test_plot_joint_bad(models, model_fit):
+    obj = getattr(models, model_fit)
+    with pytest.raises(ValueError):
+        plot_joint(obj, var_names=("mu", "tau"), kind="bad_kind")
+
+    with pytest.raises(Exception):
```
[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_plots.py]: Yes, I just didn't want to go and change code that was not broken.

Maybe add new issue and fix it in another PR

We could go through also other exceptions.

[the remark this answers is not in the document]
Comment (ahartikainen) [reply arviz/tests/test_plots.py]: https://github.com/arviz-devs/arviz/issues/517

Added issue concerning this

**Comment by ahartikainen** [comment]
so based on coveralls there are still some edge cases, but I think we can fix them in another PRs.

- io_tfp (I didn't touch)
- io_pymc3 (I didnt' touch)
- stats (our make_ufunc seems to be somehow a bit broken and one warning criteria is never met because numerical issues, but might also be due to error in algorithm, meaning wrong criteria.)
- ppcplot (I think it needs either multidimensional pp or multiple pp)
- conftest needs to simulate the issue if file is locked.

**Comment by canyon289** [comment]
I think we should merge this PR

**Comment by ahartikainen** [comment]
That sounds reasonable
